### PR TITLE
APEXCORE-496 Provide way for StatsListeners to get additional information

### PR DESCRIPTION
**Pull request by tushargosavi**

### CODE REVIEW

**Review by vrozov — CHANGES_REQUESTED** [review]
Fix checkstyle errors. Please make sure PR is ready for review.

**Comment by tushargosavi** [comment]
@vrozov I have removed checkstyle errors.

**Comment by tushargosavi** [comment]
@tweise I have rebased the PR with current master

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java [hunk 1]
```diff
@@ -1892,4 +1947,22 @@ public Integer getStreamCodecIdentifier(StreamCodec<?> streamCodecInfo)
       return null;
     }
   }
+
+  /**
+   * A interface object for the statsListener to access the DAG, it provides methods
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java]: An interface for the `{@link StatsListener}` ... Please fix spelling and formatting of other Javadoc comments as well.

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java [hunk 2]
```diff
@@ -205,6 +211,53 @@ public Response processStats(BatchedOperatorStats stats)
...
+      if (listener instanceof StatsListenerWithContext) {
+        return ((StatsListenerWithContext)listener).processStats(stats, context);
+      } else {
+        return processStats(stats);
+      }
+    }
+  }
+
+  /**
+   * Adapter for handling deprecated StatsListener. This implementation calls {@see StatsListener.processStats(BatchedOperatorStats)}
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java]: The correct formatting would be `{@link StatsListener#processStats(BatchedOperatorStats)}`

File: engine/src/test/java/com/datatorrent/stram/plan/physical/PhysicalPlanTest.java [hunk 3]
```diff
@@ -2246,4 +2253,97 @@ public void testParallelPartitionForSlidingWindow()
     PhysicalPlan plan = new PhysicalPlan(dag, new TestPlanContext());
     Assert.assertEquals("number of containers", 7, plan.getContainers().size());
   }
+
+  static class NoOpStatsListener implements StatsListener
```
Comment (tweise) [inline engine/src/test/java/com/datatorrent/stram/plan/physical/PhysicalPlanTest.java]: could this be done with mockito?

**Review by tweise — CHANGES_REQUESTED** [review]

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java (hunk 2 above)
Comment (tushargosavi) [reply engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java]: will do it

File: engine/src/test/java/com/datatorrent/stram/plan/physical/PhysicalPlanTest.java (hunk 3 above)
Comment (tushargosavi) [reply engine/src/test/java/com/datatorrent/stram/plan/physical/PhysicalPlanTest.java]: I will check with Mockito.

Comment (tushargosavi) [reply engine/src/test/java/com/datatorrent/stram/plan/physical/PhysicalPlanTest.java]: I have use Mockito to mock the interface in the tests, and updated the pull request

**Comment by tushargosavi** [comment]
@tweise  I have handled the comments.

**Review by tweise — APPROVED** [review]

**Pull request merged**